server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -219,83 +218,144 @@ public void stop()
-      }
-
-      if (segments.isEmpty()) {
-        log.warn("No segments found in the database!");
-        return false;
-      }
-
-      dbi.withHandle(
-          new HandleCallback<Void>()
-          {
-            @Override
-            public Void withHandle(Handle handle)
-            {
-              Batch batch = handle.createBatch();
+                  catch (IOException e) {
+                    throw new RuntimeException(e);
+                  }
+                }
+            )
+        )
+    );
+  }
 
-              for (DataSegment segment : segments) {
-                batch.add(
+  private VersionedIntervalTimeline<String, DataSegment> getVersionedIntervalTimeline(final String dataSource, final Collection<String> segmentIds)
+  {
+    return connector.inReadOnlyTransaction(
+        (handle, status) -> VersionedIntervalTimeline.forSegments(segmentIds.stream().map(segmentId -> {
+          try {
+            return jsonMapper.readValue(StreamSupport.stream(
+                handle.createQuery(
                     StringUtils.format(
-                        "UPDATE %s SET used=true WHERE id = '%s'",
-                        getSegmentsTable(),
-                        segment.getId()
+                        "SELECT payload FROM %1$s WHERE dataSource = :dataSource AND id = :id",
+                        getSegmentsTable()
                     )
-                );
-              }
-              batch.execute();
-
-              return null;
-            }
+                )
+                .setFetchSize(connector.getStreamingFetchSize())
+                .bind("dataSource", dataSource)
+                .bind("id", segmentId)
+                .map(ByteArrayMapper.FIRST)
+                .spliterator(), false
+            ).findFirst().orElseThrow(
+                () -> new UnknownSegmentIdException(StringUtils.format("Cannot find segment id [%s]", segmentId))
+            ), DataSegment.class);
           }
-      );
+          catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+        }).collect(Collectors.toList()))
+    );
+  }
+
+  private Stream<SegmentId> segmentIdsForInterval(
+      final VersionedIntervalTimeline<String, DataSegment> versionedIntervalTimeline,
+      final Interval interval
+  )
+  {
+    return versionedIntervalTimeline.lookup(interval).stream().flatMap(
+        objectHolder -> StreamSupport.stream(objectHolder.getObject().spliterator(), false).map(
+            dataSegmentPartitionChunk -> dataSegmentPartitionChunk.getObject().getId()
+        )
+    );
+  }
+
+  @Override
+  public boolean enableDataSource(final String dataSource)
+  {
+    try {
+      return enableSegments(dataSource, Intervals.ETERNITY) != 0;
     }
     catch (Exception e) {
       log.error(e, "Exception enabling datasource %s", dataSource);
       return false;
     }
+  }
 
-    return true;
+  @Override
+  public int enableSegments(final String dataSource, final Interval interval)
```

Comment:
Also suggest adding some comments or javadoc to these enable* methods that explain that they only enable non-overshadowed segments

Reply:
Added documentation to the Interface. Though it made more sense there as any additional implementations (if there ever are any) should behave the same way.

Reply:
`addOverlappingEnabledSegmentsToVersionIntervalTimeline` uses a potentially old, cached view of the currently enabled segments that's updated by polling (default 1 minute period), while `buildVersionedIntervalTimeline(dataSource, interval);` reads fresh from metadata. I think it would be better to do the enable based on a self-consistent view of the segments, so I'll suggest the following:

For the interval case:
1. Create an empty timeline
2. Query the metadata store for both the used status and payload of segments that overlap with the provided interval (SELECT used, payload...)
3. Go through the used/segment pairs: If the segment is used, add the segment to the timeline. If the segment is not used, check that the segments interval is contained by the provided interval. If the unused segment is contained, add the segment to the timeline, and track this segment in a separate list.
4. Go through the list of unused+contained segments in the list built in step 3, and enable the unused+contained segments for which isOvershadowed returns false. 

For the segmentId case:
1. Create an empty timeline
2. Add the provided segments to the timeline
3. Using JodaUtils.condenseIntervals(), build a condensed list of intervals from the intervals of the provided segment set
4. In a single transaction, issue a used=true + interval overlaps query for each interval in the condensed intervals list constructed in 3, and add these used+overlapping segments to the timeline
5. For each provided segment, enable it if isOvershadowed is false

Reply:
> For the segmentId case:
> 
> 1. Create an empty timeline
> 2. Add the provided segments to the timeline
This requires a query to convert the segmentId Strings to actual segments. Should this query be in the same transaction as step 4?
> 3. Using JodaUtils.condenseIntervals(), build a condensed list of intervals from the intervals of the provided segment set
> 4. In a single transaction, issue a used=true + interval overlaps query for each interval in the condensed intervals list constructed in 3 (could be a single query with an OR clause for each condensed interval), and add these used+overlapping segments to the timeline
> 5. For each provided segment, enable it if isOvershadowed is false

Reply:
I think you can actually skip the first query when adding those segments to the timeline:
- The `SegmentId` class has methods for parsing the ID string, you can grab the intervals from the parsed objects
- The timeline doesn't really need a full DataSegment object, the interval and version are what matter for determining overshadowing, and you can create a dummy partition chunk object

Reply:
You wouldn't be able to check for `UnknownSegmentIdException` case with that, but I think sacrificing that existence check to skip the query is worth it

Reply:
Sorry for the churn, let's go back to issuing a query to get the DataSegment objects for the segment ID case.

In the future, the overshadowing decision could require more information beyond just version/interval (e.g. https://github.com/apache/incubator-druid/pull/7547), and so this API would need to retrieve the full objects.